mgnify_pipelines_toolkit/analysis/shared/study_summary_generator.py

```diff
@@ -0,0 +1,263 @@
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+def parse_args():
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-i",
+        "--indir",
+        required=True,
+        type=str,
+        help="Input directory to where all the individual analyses subdirectories are if summarising, or where all the summaries are if merging",
+    )
+    parser.add_argument(
+        "-r",
+        "--runs",
+        required=False,
+        type=str,
+        help="File containing successful analyses generated by the pipeline",
+    )
+    parser.add_argument(
+        "-m",
+        "--mode",
+        required=True,
+        type=str,
+        choices=["summarise", "merge"],
+        help="Mode to either summarise analyses (summarise), or merge multiple existing summaries (merge)",
+    )
+    parser.add_argument(
+        "-p", "--prefix", required=True, type=str, help="Prefix to output"
+    )
+
+    args = parser.parse_args()
+
+    indir = args.indir
+    runs = args.runs
+    mode = args.mode
+    prefix = args.prefix
+
+    return indir, runs, mode, prefix
+
+
+def get_tax_file(run_acc, analyses_dir, db_label):
+
+    tax_file = ""
+
+    db_path = Path(f"{analyses_dir}/{run_acc}/taxonomy-summary/{db_label}")
+
+    if db_path.exists():
+        if db_label in TAXDB_LABELS:
+            tax_file = Path(
+                f"{analyses_dir}/{run_acc}/taxonomy-summary/{db_label}/{run_acc}_{db_label}.txt"
+            )
+            if not tax_file.exists():
+                logging.error(
+                    f"DB path exists but file doesn't - exiting. Path: {tax_file}"
+                )
+                exit(1)
+
+            file_size = tax_file.stat().st_size
+            if (
+                file_size == 0
+            ):  # Pipeline can generate files that are empty for ITS DBs (UNITE and ITSoneDB),
+                # so need to skip those. Should probably fix that at some point
+                tax_file = ""
+        elif db_label in ASV_TAXDB_LABELS:
+            # ASV tax files could have up to two files, one for each amplified region (maximum two from the pipeline).
+            # So will need to handle this differently to closed-reference files
+            asv_tax_files = glob.glob(
+                f"{analyses_dir}/{run_acc}/taxonomy-summary/{db_label}/*.txt"
+            )
+            asv_tax_files = [
+                Path(file) for file in asv_tax_files if "concat" not in file
+            ]  # Have to filter out concatenated file if it exists
+
+            tax_file = asv_tax_files
+
+    # output will be a Path if closed-reference file, or a list if ASV file
+    return tax_file
+
+
+def parse_one_tax_file(run_acc, tax_file):
+
+    res_df = pd.DataFrame()
+
+    with open(tax_file, "r") as fr:
+        for line in fr:
+            line = line.strip()
+            temp_lst = line.split("\t")
+            curr_count = temp_lst[0]
+            curr_tax = ";".join(temp_lst[1:])
```

Comment:
Maybe we can create a taxonomy parser of sorts, pydantic would be great fit here as we can parse/validate the content at the same time... remember _don't trust in anyone_. https://docs.pydantic.dev/latest/examples/files/#csv-files

Let's chat about this

Reply:
@mberacochea I think we could refactor out the file rules utils from https://github.com/EBI-Metagenomics/emgapi-v2/tree/main/workflows/data_io_utils/file_rules into a new package for use in this toolkit as well as in the EMG automation system. That could mean moving the file-rule things to this package, or a small new package on pypi. This would make a very nice 1 day MGnify Hackathon project.

Reply:
Have you seen this? Package that adds support for validating pandas dataframes in pydantic
https://wesselhuising.medium.com/validate-pandas-dataframes-using-pydantic-pandantic-4ec89709d5d